kohvis/src/mapping_gd.rs

```diff
@@ -0,0 +1,426 @@
+                        euclid::Rect::new(
+                            euclid::Point2D::new(x + 3., y),
+                            euclid::Size2D::new(legend as f32 - 3., total_height / steps as f32),
+                        ),
+                        color,
+                        true,
+                        1.0,
+                        false,
+                    );
+                }
+            }
+        }
+    }
+
+    /// Calculates the required columns as a vector of (index, column index).
+    fn get_columns(&self, som: &Som) -> Vec<(usize, usize)> {
+        let params = som.params();
+        let mut columns = vec![];
+        if params.layers().is_empty() || self.layers.len() == 0 {
+            columns.extend((0..som.weights().ncols()).map(|c| (c, c)));
+        } else {
+            let mut start = 0;
+            let mut index = 0;
+            for (l, layer) in params.layers().iter().enumerate() {
+                let mut contain = false;
+                for i in 0..self.layers.len() {
+                    if self.layers.get(i) == l as i32 {
+                        contain = true;
+                        break;
+                    }
+                }
+                if contain {
+                    for i in 0..layer.ncols() {
+                        columns.push((index, start + i));
+                        index += 1;
+                    }
+                }
+                start += layer.ncols();
+            }
+        };
+        columns
+    }
+    /// Calculates the optimum number of layout columns.
+    fn calc_layout_columns(
+        width: f32,
+        height: f32,
+        som_rows: usize,
+        som_cols: usize,
+        data_columns: usize,
+        heading: i32,
+        legend: i32,
+    ) -> (usize, f32) {
+        if data_columns == 1 {
+            let panel_width = width as f64 - legend as f64;
+            let panel_height = height as f64 - heading as f64;
+
+            let x_scale = panel_width / som_cols as f64;
+            let y_scale = panel_height / som_rows as f64;
+            let scale = if x_scale < y_scale { x_scale } else { y_scale };
+
+            (1, scale as f32)
+        } else {
+            (1..data_columns)
+                .map(|cols| {
+                    let layout_rows = (data_columns as f64 / cols as f64).ceil() as usize;
+                    let panel_width = (width as f64 / cols as f64) - legend as f64;
+                    let panel_height = (height as f64 / layout_rows as f64) - heading as f64;
+
+                    let x_scale = panel_width / som_cols as f64;
+                    let y_scale = panel_height / som_rows as f64;
+                    let scale = if x_scale < y_scale { x_scale } else { y_scale };
+
+                    (cols, scale as f32)
+                })
+                .max_by(|(_col1, scale1), (_col2, scale2)| scale1.partial_cmp(scale2).unwrap())
+                .unwrap()
+        }
+    }
+}
+
+impl KohonenUser2D for Mapping {
```

Comment:
IMHO, this looks dangerously like reuse-through-inheritance as traits are usually consumed by some type that is not also the implementor, e.g. in this case, `with_kohonen` could just as well be a free function with a `path` parameter. (Which would also make it possible to reuse the allocated `NodePath`.)